alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/AbstractOrientingBehavior.kt

```diff
@@ -364,6 +371,24 @@ abstract class AbstractOrientingBehavior<T, P, A, N, E, M, F>(
         pedestrian.volatileMemory.contains(area) &&
             environment.graph().outgoingEdgesOf(area).distinct().count() <= 1
 
+    /*
+     * Picks the closest door among the provided ones and sets all the state
+     * variables so as to move to the selected door.
+     * ifNull is called if no door could be found, defaults to nothing.
+     */
+    private fun moveToClosestDoor(doors: Collection<F>, ifNull: () -> Unit = {}) {
+        val currentPosition = environment.getPosition(pedestrian)
+        val closestDoor = doors
+            .map { it to computeSubdestination(it) }
+            .minBy { it.second.distanceTo(currentPosition) }
+        if (closestDoor != null) {
+            nextRoom = environment.graph().getEdgeTarget(closestDoor.first)
+            subdestination = closestDoor.second
+            targetDoor = closestDoor.first
+            state = State.MOVING_TO_DOOR
+        } else ifNull.invoke()
```

Comment:
This also looks very similar to another snippet of code I already commented, the one with `ifNoDoorIsFound`. Duplication is a source for bugs. Please check if you can extract the behaviour in a common superclass. If you can't, we'll use strategy.

Reply:
Yes there's a lot of duplication, there are two "versions" of the orienting behaviour: one deriving from `Action` and the other deriving from `Reaction`. I plan to remove the latter, but only after finding an effective method for mixing steering actions. Given that, I can maintain the duplication until that moment, or I can extract the common code in a superclass. Take into account that such superclass will be deleted in the future with every probability, so it will take some time to create it and to delete it as well. If you think it's best to collect the common code anyway, I'll define a superclass.

Reply:
Uhmm... I get it. Can't we factor it out into a strategy?
I'd like duplication issues to be solved before getting this into the mainline. We could wait for the steering mixing to be in place, though, to avoid doing things twice.

Reply:
I isolated the common code in a new class which is treated more or less like a strategy.